### PR TITLE
restyle enrich intersection points

### PR DESCRIPTION
**Pull request by barendgehrels**

### CODE REVIEW

File: include/boost/geometry/algorithms/detail/overlay/enrich_intersection_points.hpp [hunk 1]
```diff
@@ -514,18 +462,15 @@ inline void enrich_intersection_points(Turns& turns,
...
-        mit != mapped_vector.end();
-        ++mit)
+    for (auto& pair : mapped_vector)
     {
         detail::overlay::enrich_sort<Reverse1, Reverse2>(
-                    mit->second, turns,
+                    pair.second, turns,
                     geometry1, geometry2,
-                    robust_policy, strategy.side()); // TODO: pass strategy
+                    robust_policy, strategy);
```
Comment (barendgehrels) [inline include/boost/geometry/algorithms/detail/overlay/enrich_intersection_points.hpp:470]: :heavy_check_mark: done

File: include/boost/geometry/algorithms/detail/overlay/enrich_intersection_points.hpp [hunk 2]
```diff
@@ -289,92 +266,69 @@ struct enriched_map_default_include_policy
...
-    typedef typename turn_type::container_type container_type;
-    typedef typename MappedVector::mapped_type mapped_type;
-    typedef typename boost::range_value<mapped_type>::type indexed_type;
-
-    std::size_t index = 0;
-    for (typename boost::range_iterator<Turns const>::type
-            it = boost::begin(turns);
-         it != boost::end(turns);
-         ++it, ++index)
+    for_each_with_index(turns, [&](std::size_t index, auto const& turn)
```
Comment (barendgehrels) [inline include/boost/geometry/algorithms/detail/overlay/enrich_intersection_points.hpp:276]: This makes the code a lot shorter and more readable.
Better view it side-by-side.

File: include/boost/geometry/algorithms/detail/overlay/enrich_intersection_points.hpp [hunk 3]
```diff
@@ -82,24 +83,24 @@ template
...
             Geometry1 const& geometry1,
             Geometry2 const& geometry2,
             RobustPolicy const& robust_policy,
-            SideStrategy const& strategy)
+            Strategy const& strategy)
 {
-    std::sort(boost::begin(operations),
-            boost::end(operations),
+    std::sort(std::begin(operations),
+            std::end(operations),
```
Comment (vissarion) [inline include/boost/geometry/algorithms/detail/overlay/enrich_intersection_points.hpp:96]: could you please also fix indentation?

File: include/boost/geometry/algorithms/detail/overlay/enrich_intersection_points.hpp [hunk 4]
```diff
@@ -109,93 +110,83 @@ template <typename Operations, typename Turns>
 inline void enrich_assign(Operations& operations, Turns& turns,
                           bool check_turns)
 {
-    typedef typename boost::range_value<Turns>::type turn_type;
-    typedef typename turn_type::turn_operation_type op_type;
-    typedef typename boost::range_iterator<Operations>::type iterator_type;
+    if (operations.empty())
```
Comment (vissarion) [inline include/boost/geometry/algorithms/detail/overlay/enrich_intersection_points.hpp:113]: could this be a one-liner?

**Review by vissarion — APPROVED** [review]

File: include/boost/geometry/algorithms/detail/overlay/enrich_intersection_points.hpp (hunk 4 above)
Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/overlay/enrich_intersection_points.hpp:113]: I prefer to keep it like this.

File: include/boost/geometry/algorithms/detail/overlay/enrich_intersection_points.hpp [hunk 5]
```diff
@@ -289,92 +266,69 @@ struct enriched_map_default_include_policy
...
-    using op_type = typename turn_type::turn_operation_type;
-
-    typename op_type::comparable_distance_type const zero_distance = 0;
-
-    for (auto it = boost::begin(turns); it != boost::end(turns); ++it)
+    for (auto& turn : turns)
     {
-        turn_type& turn = *it;
+        using ctype = typename geometry::coordinate_type<decltype(turn.point)>::type;
+        auto distance_measure = [](auto const& a, auto const& b) -> ctype
```
Comment (barendgehrels) [inline include/boost/geometry/algorithms/detail/overlay/enrich_intersection_points.hpp:306]: This is not handled correctly by gcc up to version 7
For that reason I will undo this change to lambda

File: include/boost/geometry/algorithms/detail/overlay/enrich_intersection_points.hpp [hunk 6]
```diff
@@ -205,47 +196,37 @@ inline void enrich_assign(Operations& operations, Turns& turns,
...
-        // If it is empty, or contains one or two turns, it makes no sense
         return;
     }
 
-    // Operations is a vector of indexed_turn_operation<>
-
-    // Last index:
-    std::size_t const x = operations.size() - 1;
     bool next_phase = false;
+    std::size_t previous_index = operations.size() - 1;
```
Comment (barendgehrels) [inline include/boost/geometry/algorithms/detail/overlay/enrich_intersection_points.hpp:207]: Earlier it was slightly different, leading to a reproducable gcc-5 compilation error.
```
./boost/geometry/algorithms/detail/overlay/enrich_intersection_points.hpp:206:54: internal compiler error: in tsubst_copy, at cp/pt.c:13040
     std::size_t const last_index = operations.size() - 1;
```

This now works for all compilers AFAICS

**Comment by barendgehrels** [comment]
Now all is green, merging.

**Pull request merged**